modules/pvm-data/src/meta_store.rs

```diff
@@ -1,59 +1,63 @@
 use std::{borrow::Cow, collections::HashMap};
 
-use chrono::{DateTime, Utc};
+use super::{node_types::ConcreteType, ID};
 
-type MetaEntry = (String, u8);
+type MetaEntry = (String, ID);
 
 type MetaType = (bool, Vec<MetaEntry>);
 
 #[derive(Clone, Deserialize, Debug, Default, Serialize)]
 pub struct MetaStore {
     entries: HashMap<Cow<'static, str>, MetaType>,
-    time_list: Vec<DateTime<Utc>>,
 }
 
 impl MetaStore {
     pub fn new() -> Self {
         MetaStore {
             entries: HashMap::new(),
-            time_list: Vec::new(),
         }
     }
 
-    pub fn snapshot(&self, time: &DateTime<Utc>) -> Self {
+    pub fn from_map(
+        src: HashMap<&'static str, String>,
+        ctx: ID,
+        ty: &'static ConcreteType,
+    ) -> Self {
+        MetaStore {
+            entries: src
+                .into_iter()
+                .map(|(k, v)| (k.into(), (ty.props[k], vec![(v, ctx)])))
+                .collect(),
+        }
+    }
+
+    pub fn snapshot(&self, ctx: ID) -> Self {
         let entries: HashMap<Cow<'static, str>, MetaType> = self
             .entries
             .iter()
             .filter_map(|(n, (h, v))| {
                 if *h {
                     let (v_last, _) = v.last().unwrap();
-                    Some((n.clone(), (true, vec![(v_last.clone(), 0)])))
+                    Some((n.clone(), (true, vec![(v_last.clone(), ctx)])))
                 } else {
                     None
                 }
             })
             .collect();
-        let time_list = {
-            if !entries.is_empty() {
-                vec![*time]
-            } else {
-                vec![]
-            }
-        };
-        MetaStore { entries, time_list }
+        MetaStore { entries }
     }
 
     pub fn merge(&mut self, other: &MetaStore) {
         for (k, v, t, h) in other.iter() {
```

Comment:
it would improve readability if the letters had some relation to what they represent.
k and v are clearly key and value, I assume t was from "time" but now is "context" so
would suggest changing this to c or "ctx".

Reply:
Corrections made.